src/page/page_cache.rs

```diff
@@ -179,54 +178,102 @@ impl<M> PageCache<LockFreeLog, M>
+                match entry {
+                    CacheEntry::Resident(_, ref lid) => {
+                        new_stack.push(CacheEntry::PartialFlush(*lid));
+                    }
+                    CacheEntry::Flush(_) => panic!("got Flush in middle of stack..."),
+                    CacheEntry::PartialFlush(lid) => {
+                        new_stack.push(CacheEntry::PartialFlush(lid));
+                    }
+                }
+            }
+            new_stack.push(last.unwrap());
+            let node = node_from_frag_vec(new_stack);
+            let res = unsafe { (*stack_ptr).cas(head, node) };
+            if res.is_ok() {
+                self.lru.page_out_succeeded(pid);
+            } else {
+                // if this failed, it's because something wrote to the page in the mean time
+            }
+        }
     }
 
     fn page_in(&self,
                pid: PageID,
-               head: *const stack::Node<CacheEntry<M>>)
+               mut head: *const stack::Node<CacheEntry<M>>,
+               stack_ptr: *const stack::Stack<CacheEntry<M>>)
                -> (M::MaterializedPage, *const stack::Node<CacheEntry<M>>) {
         let stack_iter = StackIter::from_ptr(head);
         let mut cache_entries: Vec<CacheEntry<M>> = stack_iter.map(|ptr| (*ptr).clone()).collect();
 
-        // "fill-in" any flushes
-        cache_entries.par_iter_mut()
-            .for_each(|ref mut ce| {
-                let (pp, lid) = match *ce {
-                    &mut CacheEntry::Resident(ref pp, ref lid) => {
-                        if let &Some(l) = lid {
-                            self.lru.accessed(l);
-                        }
-                        (pp.clone(), lid.clone())
-                    }
-                    &mut CacheEntry::Flush(lid) => (self.pull(lid).unwrap(), Some(lid)),
-                    &mut CacheEntry::PartialFlush(lid) => (self.pull(lid).unwrap(), Some(lid)),
-                };
-                **ce = CacheEntry::Resident(pp, lid);
-            });
+        // read items off of disk in parallel if anything isn't already resident
+        let contains_non_resident = cache_entries.iter().any(|ce| match *ce {
+            CacheEntry::Resident(_, _) => false,
+            _ => true,
+        });
+        if contains_non_resident {
+            cache_entries.par_iter_mut()
+                .for_each(|ref mut ce| {
+                    let (pp, lid) = match *ce {
+                        &mut CacheEntry::Resident(ref pp, ref lid) => (pp.clone(), lid.clone()),
+                        &mut CacheEntry::Flush(lid) => (self.pull(lid).unwrap(), lid),
+                        &mut CacheEntry::PartialFlush(lid) => (self.pull(lid).unwrap(), lid),
+                    };
+                    **ce = CacheEntry::Resident(pp, lid);
+                });
+
+            let node = node_from_frag_vec(cache_entries.clone());
+            let res = unsafe { (*stack_ptr).cas(head, node) };
```

Comment:
here's the bug I realized existed while we were talking, I had forgotten to actually swap the stack after reading the values off of disk. Now it only reads these off disk when they are not actually present (`contains_non_resident` above)